File: src/io/chanse/hackerrank/practice/interviewprepkit/arrays/LeftRotation.kt
package io.chanse.hackerrank.practice.interviewprepkit.arrays

import java.util.*

// https://www.hackerrank.com/challenges/ctci-array-left-rotation/problem?h_l=interview&playlist_slugs%5B%5D%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D%5B%5D=arrays
fun rotLeft(a: Array<Int>, d: Int): Array<Int> {
    val arrSize = a.size
    val diff = arrSize - d
    val output = Array(arrSize) { 0 }
    a.forEachIndexed { index, i ->
        output[(index + diff) % arrSize] = i
    }
    return output
}

fun main() {
    val scan = Scanner(System.`in`)

    val nd = scan.nextLine().split(" ")

    val n = nd[0].trim().toInt()
    val d = nd[1].trim().toInt()
    val a = scan.nextLine()
        .split(" ")
        .map{ it.trim().toInt() }
        .toTypedArray()

    val result = rotLeft(a, d)
    println(result.joinToString(" "))
}

File: src/io/chanse/hackerrank/practice/interviewprepkit/arrays/LeftRotation.py


# https://www.hackerrank.com/challenges/ctci-array-left-rotation/problem?h_l=interview&playlist_slugs%5B%5D%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D%5B%5D=arrays
def rot_left(a, d):
    arr_size = len(a)
    diff = arr_size - d
    output = [0] * arr_size

    for index, value in enumerate(a):
        output[(index + diff) % arr_size] = value
    return output


if __name__ == '__main__':
    # fptr = open(os.environ['OUTPUT_PATH'], 'w')

    nd = input().split()
    n = int(nd[0])
    d = int(nd[1])

    a = list(map(int, input().rstrip().split()))

    result = rot_left(a, d)
    print(result)

    # fptr.write(' '.join(map(str, result)))
    # fptr.write('\n')
    # fptr.close()
